### ISSUE TITLE
Spkl: remove non-existing plugins (or all)

### ISSUE THREAD

**Issue by KimBon** [scottdurow/SparkleXrm#197]
Hi Scott

thanks for this great tool! At one of our projects, we are looking to remove all plugin steps on our build environment, and after this, re-register plugins based on config.

We will probably use spkl for the second part of the requirement. But would it not be an option an option to have a setting to remove existing plugin steps, or to remove only the ones not mentioned via spkl registration? If I understood right from your code, you only add or replace plugin steps. In a CI setup, would it not be advisable to remove them first?

Thanks.

**Comment by scottdurow** [comment]
I agree that we need a way of removing registrations (I have a 'remove' option on the step registrations but yet to be implemented!) I'm not so sure about removing everything and re-adding because this would give new GUIDs to the registrations which could make solution exports unnecessarily updated and possibly lead to duplicate steps being created in the target environment?

**Comment by KimBon** [comment]
What we would like to avoid is our plugin registration having duplicate steps or unnecessary steps after deployment. That's why we were thinking on removing all. Can you explain me a bit more what's exactly the down-side of removing them all? And how would removal for a single registration work as the class would be removed?

**Comment by mattp65** [comment]
I can see a need for a "remove" option, but being used more in Release than in Build. For envs that aren't using Managed solutions, the ability to have a Remove can be important and the Solution Import afterwards would put everything back.

Also, I could see the update process for the Steps and Images being more of a "sync" process so that if an individual step or image is removed from the Metadata tags, the utility removes them during the deploy. Would require pulling the existing records into a collection, then processing them to remove any no longer specified (maybe remove from the collection as they are processed initially). This would address the issue of creating new GUIDs, but also allow the assembly metadata to be the source of truth for the steps/images without a need for manual intervention. Psuedo code:

```
Get CurList
For Each entry in the Assembly Metadata
    Process the insert/Update
    Remove from CurList
For Each entry in CurList
    Delete
```

If that was done within the Images & Step segments, it would keep the registration in sync with the original source.

**Comment by scottdurow** [comment]
Agreed - sync might be a better option compared to remove.

**Comment by mattp65** [comment]
I put in a Pull request that does the "sync" on the Steps/Images based on the Metadata in the updated DLL. 

"Removing" an entire method via this mechanism isn't really an option since the initial update errors out if a method is removed. Remove could still be handled with a new Task (Delete-Plugin, Delete-PluginType) that is placed in the JSON to run before processing of of the updated plugin.

**Comment by maheshghori** [comment]
Hi Scott,

This tool helps a lot for DevOps. Thanks a lot. I have one query regarding using it. I need to remove one plugin step. Is it possible using CrmPluginRegistrationAttribute ?

Thanks,
Mahesh Ghori
